velox/functions/sparksql/SIMDJsonFunctions.h

```diff
@@ -0,0 +1,192 @@
+          }
+          case ondemand::number_type::signed_integer: {
+            int64_t numberResult;
+            error = rawResult.get_int64().get(numberResult);
+            if (!error) {
+              ss << numberResult;
+              result.append(ss.str());
+            }
+            return error;
+          }
+          case ondemand::number_type::floating_point_number: {
+            double numberResult;
+            error = rawResult.get_double().get(numberResult);
+            if (!error) {
+              ss << numberResult;
+              result.append(ss.str());
+            }
+            return error;
+          }
+          default:
+            VELOX_UNREACHABLE();
+        }
+      }
+      case ondemand::json_type::boolean: {
+        bool boolResult = false;
+        rawResult.get_bool().get(boolResult);
+        if (boolResult) {
+          result.append("true");
+        } else {
+          result.append("false");
+        }
+        return SUCCESS;
+      }
+      case ondemand::json_type::string: {
+        std::string_view stringResult;
+        error = rawResult.get_string().get(stringResult);
+        result.append(stringResult);
+        return error;
+      }
+      case ondemand::json_type::object: {
+        // For nested case, e.g., for "{"my": {"hello": 10}}", "$.my" will
+        // return an object type.
+        ss << rawResult;
+        result.append(ss.str());
+        return SUCCESS;
+      }
+      case ondemand::json_type::array: {
+        ss << rawResult;
+        result.append(ss.str());
+        return SUCCESS;
+      }
+      default: {
+        return UNSUPPORTED_ARCHITECTURE;
+      }
+    }
+  }
+
+  // This is a simple validation by checking whether the obtained result is
+  // followed by valid char. Because ondemand parsing we are using ignores json
+  // format validation for characters following the current parsing position.
+  bool isValidEndingCharacter(const char* currentPos) {
+    char endingChar = *currentPos;
+    if (endingChar == ',') {
+      return true;
+    } else if (endingChar == '}') {
+      return true;
+    } else if (endingChar == ']') {
+      return true;
+    } else if (
+        endingChar == ' ' || endingChar == '\r' || endingChar == '\n' ||
+        endingChar == '\t') {
+      // These chars can be prior to a valid ending char.
+      return isValidEndingCharacter(currentPos++);
+    } else {
+      return false;
+    }
+  }
+
+  FOLLY_ALWAYS_INLINE bool call(
+      out_type<Varchar>& result,
+      const arg_type<Varchar>& json,
+      const arg_type<Varchar>& jsonPath) {
+    ParserContext ctx(json.data(), json.size());
+    try {
+      ctx.parseDocument();
+      simdjson_result<ondemand::value> rawResult;
+      if (formattedJsonPath_.has_value()) {
+        rawResult = ctx.jsonDoc.at_pointer(formattedJsonPath_.value().data());
+      } else {
+        rawResult =
+            ctx.jsonDoc.at_pointer(getFormattedJsonPath(jsonPath).data());
+      }
```

Comment:
simdjson_result<ondemand::value> rawResult = formattedJsonPath_.has_value() ? one : the other